src/fu-hwids.c

```diff
@@ -236,81 +237,121 @@ fu_hwids_get_guid (FuHwids *self, const gchar *keys, GError **error)
+				    GError **error)
+{
+	GBytes *data;
+	const guint8 *data_raw;
+	gsize data_sz = 0;
+	data = fu_smbios_get_data (smbios, type, error);
+	if (data == NULL)
+		return NULL;
+	data_raw = g_bytes_get_data (data, &data_sz);
+	if (offset >= data_sz) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "offset bigger than data");
+		return NULL;
+	}
+	return g_strdup_printf ("%u", data_raw[offset]);
+}
+
 /**
  * fu_hwids_setup:
  * @self: A #FuHwids
- * @sysfsdir: The sysfs directory, or %NULL for the default
+ * @smbios: A #FuSmbios
  * @error: A #GError or %NULL
  *
  * Reads all the SMBIOS values from the hardware.
  *
  * Returns: %TRUE for success
  **/
 gboolean
-fu_hwids_setup (FuHwids *self, const gchar *sysfsdir, GError **error)
+fu_hwids_setup (FuHwids *self, FuSmbios *smbios, GError **error)
 {
 	struct {
-		const gchar *key;
-		const gchar *value;
-	} sysfsfile[] = {
-		{ FU_HWIDS_KEY_MANUFACTURER,		"sys_vendor" },
-		{ FU_HWIDS_KEY_ENCLOSURE_KIND,		"chassis_type" },
-		{ FU_HWIDS_KEY_FAMILY,			"product_family" },
-		{ FU_HWIDS_KEY_PRODUCT_NAME,		"product_name" },
-		{ FU_HWIDS_KEY_PRODUCT_SKU,		"product_sku" },
-		{ FU_HWIDS_KEY_BIOS_VENDOR,		"bios_vendor" },
-		{ FU_HWIDS_KEY_BIOS_VERSION,		"bios_version" },
-		{ FU_HWIDS_KEY_BIOS_MAJOR_RELEASE,	"bios_major_release" },
-		{ FU_HWIDS_KEY_BIOS_MINOR_RELEASE,	"bios_minor_release" },
-		{ FU_HWIDS_KEY_BASEBOARD_MANUFACTURER,	"board_vendor" },
-		{ FU_HWIDS_KEY_BASEBOARD_PRODUCT,	"board_name" },
-		{ NULL, NULL }
+		const gchar		*key;
+		guint8			 type;
+		guint8			 offset;
+		FuHwidsConvertFunc	 func;
+	} map[] = {
+		{ FU_HWIDS_KEY_MANUFACTURER,		FU_SMBIOS_STRUCTURE_TYPE_SYSTEM, 0x04,
+							fu_hwids_convert_string_table_cb },
+		{ FU_HWIDS_KEY_ENCLOSURE_KIND,		FU_SMBIOS_STRUCTURE_TYPE_CHASSIS, 0x05,
+							fu_hwids_convert_base10_integer_cb },
+		{ FU_HWIDS_KEY_FAMILY,			FU_SMBIOS_STRUCTURE_TYPE_SYSTEM, 0x1a,
+							fu_hwids_convert_string_table_cb },
+		{ FU_HWIDS_KEY_PRODUCT_NAME,		FU_SMBIOS_STRUCTURE_TYPE_SYSTEM, 0x05,
+							fu_hwids_convert_string_table_cb },
+		{ FU_HWIDS_KEY_PRODUCT_SKU,		FU_SMBIOS_STRUCTURE_TYPE_SYSTEM, 0x19,
+							fu_hwids_convert_string_table_cb },
+		{ FU_HWIDS_KEY_BIOS_VENDOR,		FU_SMBIOS_STRUCTURE_TYPE_BIOS, 0x04,
+							fu_hwids_convert_string_table_cb },
+		{ FU_HWIDS_KEY_BIOS_VERSION,		FU_SMBIOS_STRUCTURE_TYPE_BIOS, 0x05,
+							fu_hwids_convert_string_table_cb },
+		{ FU_HWIDS_KEY_BIOS_MAJOR_RELEASE,	FU_SMBIOS_STRUCTURE_TYPE_BIOS, 0x14,
+							fu_hwids_convert_base10_integer_cb },
+		{ FU_HWIDS_KEY_BIOS_MINOR_RELEASE,	FU_SMBIOS_STRUCTURE_TYPE_BIOS, 0x15,
+							fu_hwids_convert_base10_integer_cb },
+		{ FU_HWIDS_KEY_BASEBOARD_MANUFACTURER,	FU_SMBIOS_STRUCTURE_TYPE_BASEBOARD, 0x04,
+							fu_hwids_convert_string_table_cb },
+		{ FU_HWIDS_KEY_BASEBOARD_PRODUCT,	FU_SMBIOS_STRUCTURE_TYPE_BASEBOARD, 0x05,
+							fu_hwids_convert_string_table_cb },
+		{ NULL, 0x00, 0x00, NULL }
```

Comment:
I do think it would be useful to parse OEM strings and include them too if possible.

Reply:
OEM strings are not part of the hardware ID: https://docs.microsoft.com/en-gb/windows-hardware/drivers/install/specifying-hardware-ids-for-a-computer -- I'm hesitant to start defining my own, and would rather just use the set defined in Windows 10.

Reply:
Right.  So maybe this is just a useful extension for plugins to use but not to create HWID's from.

At least for Dell the reason it's useful is there are instances the machine has been rebranded and the only 100% accurate way to identify a machine are through OEM strings.  Since I had to tear that out until I'm sure what's wrong with libsmbios on a few machines, this isn't too important right now though.

Reply:
Can't you just call `fu_plugin_get_smbios_string()` from the Dell plugin?